runtime/codert_vm/cnathelp.cpp

```diff
@@ -59,6 +59,37 @@ samePCs(void *pc1, void *pc2)
 #define samePCs(pc1, pc2) (MASK_PC(pc1) == MASK_PC(pc2))
 #endif /* J9ZOS390 && !J9VM_ENV_DATA64 */
 
+/**
+ * Fix the java and decompilation stacks for cases where exceptions can be
+ * thrown from insde a JIT synthetic exception handler. There must be a
+ * resolve frame on the top of the java stack with a valid JIT PC.
+ *
+ * @param currentThread[in] the current J9VMThread
+ */
+static void
+fixStackForSyntheticHandler(J9VMThread *currentThread)
+{
+	/* If the top decompilation record is for the current compiled frame, re-link the resolve frame
+	 * and decompilation record to keep the stack walkable.
+	 */
+	J9JITDecompilationInfo *decomp = currentThread->decompilationStack;
+	if (NULL != decomp) {
+		J9SFJITResolveFrame *resolveFrame = (J9SFJITResolveFrame*)currentThread->sp;
+		void *jitPC = resolveFrame->returnAddress;
+		J9JITExceptionTable *metaData = jitGetExceptionTableFromPC(currentThread, (UDATA)jitPC);
+		Assert_CodertVM_false(NULL == metaData);
+		UDATA *oldSP = (UDATA*)(resolveFrame + 1);
+		UDATA *bp = oldSP + getJitTotalFrameSize(metaData);
+		if (bp == decomp->bp) {
+			/* Use a value which is guaranteed to fail metadata lookup. This makes
+			 * the stack walker get the PC from the decompilation stack.
+			 */
+			resolveFrame->returnAddress = NULL;
```

Comment:
I see the code to check the decomp stack in jswalk[1] but shouldn't this be modifying the PC here and was having trouble finding the code that treated the `returnAddress` as the `walkState->pc`.  Found it now [2].

Do we use this pattern elsewhere as well?  Or is this the first time we're taking advantage of the way jswalk works in this way?

[1] https://github.com/eclipse/openj9/blob/957a07826573fa58672a2e5a6735f7051b91be63/runtime/codert_vm/jswalk.c#L1418-L1424

[2] https://github.com/eclipse/openj9/blob/957a07826573fa58672a2e5a6735f7051b91be63/runtime/codert_vm/jswalk.c#L341

Reply:
The code for reporting exception catch does a similarly nasty fixup:

https://github.com/eclipse/openj9/blob/957a07826573fa58672a2e5a6735f7051b91be63/runtime/codert_vm/decomp.cpp#L2459-L2466